File: TinyScript/java/src/main/java/paser/util/ParserUtils.java
package paser.util;

import jdk.jshell.spi.ExecutionControl;
import paser.ast.ASTNode;
import paser.ast.Factor;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

/**
 * @author MFine
 * @version 1.0
 * @date 2021/10/12 0:39
 **/
public class ParserUtils {

    public static String toPostfixExpression(ASTNode node) throws ExecutionControl.NotImplementedException {
        if (node instanceof Factor) {
            return node.getLexeme().getValue();
        }
        List<String> ptr = new ArrayList<>();
        for (ASTNode child : node.getChildren()) {
            String s = toPostfixExpression(child);
            ptr.add(s);
        }
        String s = node.getLexeme() != null ? node.getLexeme().getValue() : "";
        return String.join(" ", ptr) + s;
    }

    public static String toBFSString(ASTNode root, int max) {
        var queue = new LinkedList<ASTNode>();
        var list = new ArrayList<String>();
        queue.add(root);
        int c = 0;
        while (queue.size() > 0 && c++ < max) {
            var node = queue.poll();
            list.add(node.getLabel());
            for (ASTNode child : node.getChildren()) {
                queue.add(child);
            }
        }
        return String.join(" ", list);
    }
}

File: TinyScript/js/src/main/js/parser/util/ParserUtils.ts
import ASTNode from "../ast/ASTNode";
import ASTNodeTypes from "../ast/ASTNodeTypes";
import LinkedList from "ts-linked-list";

export default class ParserUtils {
    static toPostfixExpression(node: ASTNode | null): string | undefined {
        if (node == null) {
            return "";
        }
        let leftStr = "";
        let right = "";
        switch (node.type) {
            case ASTNodeTypes.DECLARE_STMT:
            case ASTNodeTypes.ASSIGN_STMT:
            case ASTNodeTypes.BINARY_EXPR:
                leftStr = ParserUtils.toPostfixExpression(node.children[0]) ?? "";
                right = ParserUtils.toPostfixExpression(node.children[1]) ?? "";
                return leftStr + " " + right + " " + node.lexeme?.value ?? " ";
            case ASTNodeTypes.VARIABLE:
            case ASTNodeTypes.SCALAR:
                return node.lexeme?.value ?? "";
            default:
                break;
        }
    }

    static toBFSString(root: ASTNode, max: number): string {
        const queue: Array<ASTNode> = [];
        const list: Array<string> = [];
        queue.push(root);
        let c = 0;
        while (queue.length > 0 && c++ < max) {
            const node = queue.shift();
            if (node && node.label) {
                list.push(node.label);
                queue.push(...node.children)
            }
        }
        return list.join(" ");
    }
}
